File: abstractions/java/lib/src/main/java/com/microsoft/kiota/RequestInfo.java
package com.microsoft.kiota;

import java.net.URI;
import java.net.URISyntaxException;
import java.io.IOException;
import java.io.InputStream;
import java.util.Arrays;
import java.util.Collection;
import java.util.HashMap;
import java.util.Objects;
import java.util.function.Function;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

import com.microsoft.kiota.serialization.Parsable;
import com.microsoft.kiota.serialization.SerializationWriter;
import com.microsoft.kiota.HttpCore;

/** This class represents an abstract HTTP request. */
public class RequestInfo {
    /** The URI of the request. */
    @Nullable
    public URI uri;
    /**
     * Sets the URI of the request.
     * @param currentPath the current path (scheme, host, port, path, query parameters) of the request.
     * @param pathSegment the segment to append to the current path.
     * @param isRawUrl whether the path segment is a raw url. When true, the segment is not happened and the current path is parsed for query parameters.
     */
    public void setUri(@Nullable final String currentPath, @Nullable final String pathSegment, final boolean isRawUrl) {
        if (isRawUrl) {
            if(currentPath == null || currentPath.isEmpty()) {
                throw new IllegalArgumentException("currentPath cannot be null or empty");
            }
            final var questionMarkSplat = currentPath.split("?");
            final var schemeHostAndPath = questionMarkSplat[0];
            this.setUriFromString(schemeHostAndPath);
            if (questionMarkSplat.length > 1) {
                final var queryString = questionMarkSplat[1];
                final var rawQueryParameters = queryString.split("&");
                for (var queryParameter : rawQueryParameters) {
                    final var queryParameterNameValue = queryParameter.split("=");
                    if (!queryParameterNameValue[0].isEmpty()) {
                        this.queryParameters.put(queryParameterNameValue[0], queryParameterNameValue.length > 1 ? queryParameterNameValue[1] : null);
                    }
                }
            }
        } else {
            this.setUriFromString(currentPath + pathSegment);
        }
    }
    private void setUriFromString(final String uriString) {
        try {
            this.uri = new URI(uriString);
        } catch (final URISyntaxException e) {
            throw new RuntimeException(e);
        }
    }
    /** The HTTP method for the request */
    @Nullable
    public HttpMethod httpMethod;
    /** The Query Parameters of the request. */
    @Nonnull
    public HashMap<String, Object> queryParameters = new HashMap<>(); //TODO case insensitive
    /** The Request Headers. */
    @Nonnull
    public HashMap<String, String> headers = new HashMap<>(); // TODO case insensitive
    /** The Request Body. */
    @Nullable
    public InputStream content;
    private HashMap<String, MiddlewareOption> _middlewareOptions = new HashMap<>();
    /**
     * Gets the middleware options for this request. Options are unique by type. If an option of the same type is added twice, the last one wins.
     * @return the middleware options for this request.
     */
    public Collection<MiddlewareOption> getMiddlewareOptions() { return _middlewareOptions.values(); }
    /**
     * Adds a middleware option to this request.
     * @param option the middleware option to add.
     */
    public void addMiddlewareOptions(@Nullable final MiddlewareOption... options) { 
        if(options == null || options.length == 0) return;
        for(final MiddlewareOption option : options) {
            _middlewareOptions.put(option.getClass().getCanonicalName(), option);
        }
    }
    /**
     * Removes a middleware option from this request.
     * @param option the middleware option to remove.
     */
    public void removeMiddlewareOptions(@Nullable final MiddlewareOption... options) {
        if(options == null || options.length == 0) return;
        for(final MiddlewareOption option : options) {
            _middlewareOptions.remove(option.getClass().getCanonicalName());
        }
    }
    private static String binaryContentType = "application/octet-stream";
    private static String contentTypeHeader = "Content-Type";
    /**
     * Sets the request body to be a binary stream.
     * @param value the binary stream
     */
    public void setStreamContent(@Nonnull final InputStream value) {
        Objects.requireNonNull(value);
        this.content = value;
        headers.put(contentTypeHeader, binaryContentType);
    }
    /**
     * Sets the request body from a model with the specified content type.
     * @param values the models.
     * @param contentType the content type.
     * @param httpCore The core service to get the serialization writer from.
     * @param <T> the model type.
     */
    public <T extends Parsable> void setContentFromParsable(@Nonnull final HttpCore httpCore, @Nonnull final String contentType, @Nonnull final T... values) {
        Objects.requireNonNull(httpCore);
        Objects.requireNonNull(values);
        Objects.requireNonNull(contentType);
        if(values.length == 0) throw new RuntimeException("values cannot be empty");

        try(final SerializationWriter writer = httpCore.getSerializationWriterFactory().getSerializationWriter(contentType)) {
            headers.put(contentTypeHeader, contentType);
            if(values.length == 1) 
                writer.writeObjectValue(null, values[0]);
            else
                writer.writeCollectionOfObjectValues(null, Arrays.asList(values));
            this.content = writer.getSerializedContent();
        } catch (IOException ex) {
            throw new RuntimeException("could not serialize payload", ex);
        }
    }
}

File: abstractions/go/request_info.go
package abstractions

import (
	"errors"
	"reflect"
	"strings"

	"net/url"
	u "net/url"

	s "github.com/microsoft/kiota/abstractions/go/serialization"
)

/* This type represents an abstract HTTP request. */
type RequestInfo struct {
	Method          HttpMethod
	URI             u.URL
	Headers         map[string]string
	QueryParameters map[string]string
	Content         []byte
	options         map[string]MiddlewareOption
}

func NewRequestInfo() *RequestInfo {
	return &RequestInfo{
		URI:             u.URL{},
		Headers:         make(map[string]string),
		QueryParameters: make(map[string]string),
		options:         make(map[string]MiddlewareOption),
	}
}

func (request *RequestInfo) SetUri(currentPath string, pathSegment string, isRawUrl bool) error {
	if isRawUrl {
		if currentPath == "" {
			return errors.New("current path cannot be empty")
		}
		questionMarkSplat := strings.Split(currentPath, "?")
		schemeHostAndPath := questionMarkSplat[0]
		uri, err := url.Parse(schemeHostAndPath)
		if err != nil {
			return err
		}
		request.URI = *uri
		if len(questionMarkSplat) > 1 {
			queryParameters := questionMarkSplat[1]
			for _, queryParameter := range strings.Split(queryParameters, "&") {
				keyValue := strings.Split(queryParameter, "=")
				if len(keyValue) == 2 {
					request.QueryParameters[keyValue[0]] = keyValue[1]
				} else if len(keyValue) == 1 {
					request.QueryParameters[keyValue[0]] = ""
				}
			}
		}
	} else {
		uri, err := url.Parse(currentPath + pathSegment)
		if err != nil {
			return err
		}
		request.URI = *uri
	}
	return nil
}

func (request *RequestInfo) AddMiddlewareOptions(options ...MiddlewareOption) error {
	if options == nil {
		return errors.New("MiddlewareOptions cannot be nil")
	}
	if request.options == nil {
		request.options = make(map[string]MiddlewareOption, len(options))
	}
	for _, option := range options {
		tp := reflect.TypeOf(option)
		name := tp.Name()
		request.options[name] = option
	}
	return nil
}

func (request *RequestInfo) GetMiddlewareOptions() []MiddlewareOption {
	if request.options == nil {
		return []MiddlewareOption{}
	}
	result := make([]MiddlewareOption, len(request.options))
	for _, option := range request.options {
		result = append(result, option)
	}
	return result
}

const contentTypeHeader = "Content-Type"
const binaryContentType = "application/octet-steam"

func (request *RequestInfo) SetStreamContent(content []byte) {
	request.Content = content
	request.Headers[contentTypeHeader] = binaryContentType
}
func (request *RequestInfo) SetContentFromParsable(coreService HttpCore, contentType string, items ...s.Parsable) error {
	if contentType == "" {
		return errors.New("content type cannot be empty")
	} else if coreService == nil {
		return errors.New("coreService cannot be nil")
	} else if items == nil || len(items) == 0 {
		return errors.New("items cannot be nil or empty")
	}
	factory, err := coreService.GetSerializationWriterFactory()
	if err != nil {
		return err
	} else if factory == nil {
		return errors.New("factory cannot be nil")
	}
	writer, err := factory.GetSerializationWriter(contentType)
	if err != nil {
		return err
	} else if writer == nil {
		return errors.New("writer cannot be nil")
	}
	defer writer.Close()
	var writeErr error
	if len(items) == 1 {
		writeErr = writer.WriteObjectValue("", items[0])
	} else {
		writeErr = writer.WriteCollectionOfObjectValues("", items)
	}
	if writeErr != nil {
		return writeErr
	}
	content, err := writer.GetSerializedContent()
	if err != nil {
		return err
	} else if content == nil {
		return errors.New("content cannot be nil")
	}
	request.Content = content
	request.Headers[contentTypeHeader] = contentType
	return nil
}
